server/src/main/java/com/cloud/server/StatsCollector.java

```diff
@@ -497,86 +572,35 @@ protected void runInContext() {
-                                    metrics.put(externalStatsPrefix + "cloudstack.stats.instances." + vmName + ".disk.read_iops", statsForCurrentIteration.getDiskReadIOs());
-                                    metrics.put(externalStatsPrefix + "cloudstack.stats.instances." + vmName + ".memory.total_kbs", statsForCurrentIteration.getMemoryKBs());
-                                    metrics.put(externalStatsPrefix + "cloudstack.stats.instances." + vmName + ".memory.internalfree_kbs", statsForCurrentIteration.getIntFreeMemoryKBs());
-                                    metrics.put(externalStatsPrefix + "cloudstack.stats.instances." + vmName + ".memory.target_kbs", statsForCurrentIteration.getTargetMemoryKBs());
+                                storeVirtualMachineStatsInMemory(statsForCurrentIteration);
 
+                                if (externalStatsType == ExternalStatsProtocol.GRAPHITE) {
+                                    prepareVmMetricsForGraphite(metrics, statsForCurrentIteration);
+                                } else {
+                                    metrics.put(statsForCurrentIteration.getVmId(), statsForCurrentIteration);
                                 }
-
                             }
 
-                            /**
-                             * Send the metrics to a external stats collector
-                             * We send it on a per-host basis to prevent that we flood the host
-                             * Currently only Graphite is supported
-                             */
                             if (!metrics.isEmpty()) {
-                                if (externalStatsType != null && externalStatsType == ExternalStatsProtocol.GRAPHITE) {
-
-                                    if (externalStatsPort == -1) {
-                                        externalStatsPort = 2003;
-                                    }
-
-                                    s_logger.debug("Sending VmStats of host " + host.getId() + " to Graphite host " + externalStatsHost + ":" + externalStatsPort);
-
-                                    try {
-                                        GraphiteClient g = new GraphiteClient(externalStatsHost, externalStatsPort);
-                                        g.sendMetrics(metrics);
-                                    } catch (GraphiteException e) {
-                                        s_logger.debug("Failed sending VmStats to Graphite host " + externalStatsHost + ":" + externalStatsPort + ": " + e.getMessage());
-                                    }
-
-                                    metrics.clear();
+                                if (externalStatsType == ExternalStatsProtocol.GRAPHITE) {
+                                    sendVmMetricsToGraphiteHost(metrics, host);
+                                } else if (externalStatsType == ExternalStatsProtocol.INFLUXDB) {
+                                    sendMetricsToInfluxdb(metrics);
```

Comment:
@DaanHoogland sorry, the lines 600 - 603 have changed, now are 596 and 598. 
I was referring to the following mthods `sendVmMetricsToGraphiteHost(metrics, host);` and `sendMetricsToInfluxdb(metrics);`

I had two options: (i) keep the Graphite metrics Map as it was, and create a new map for the influxdb (for instance metricsInfluxDB):
```
if (externalStatsType == ExternalStatsProtocol.GRAPHITE) {
    sendVmMetricsToGraphiteHost(metrics, host);
} else if (externalStatsType == ExternalStatsProtocol.INFLUXDB) {
    sendMetricsToInfluxdb(metricsInfluxDB);
```
Or (ii) change the metrics map from `Map<String, Integer>` to a generic `Map<Object, Object>`, which implicates on changing the GraphiteClient.java sendMetrics method, and reuse the `metrics` for InfluxDB and Graphite.